### DISCUSSION TITLE
Python unit test

### DISCUSSION THREAD

**Discussion by BECATRUE** [snu-quiqcl/qiwis#83] [Ideas]
As the project grew bigger, using test tool became necessary.

I think we need to complete unit tests before cooperating with Ewha Womans University, then what about your opinions?

**Comment by BECATRUE** [comment]
Please leave your opinion, freely!

@kangz12345 @Ahrrri

**Reply by kangz12345** [reply]
I agree, I think we should to unit tests.

**Reply by BECATRUE** [reply]
Okay! Then, I'll set up the initial setting.

**Comment by BECATRUE** [comment]
Now, let's talk about testing system.

I suggest an idea:
1. Use `unittest`. See also: [unittest](https://docs.python.org/3/library/unittest.html) 
 
2. In each directory, there would be one test file, i.e. `tests.py`.
For example, `swift` directory tree is as below:  
<img width="10%" alt="image" src="https://user-images.githubusercontent.com/76851886/225791903-94d0fdac-3213-418a-bf28-617d53ae46b4.png">

3. In `tests.py`, there would be several test classes and each class represents a file in the directory.
Concretely, in the above directory, there are three classes; `AppTest`, `BusTest`, and `SwiftTest`.
Of course, each class is a subclass of `unittest.TestCase`.

Please, leave your idea freely!

**Reply by BECATRUE** [reply]
@kangz12345 @Ahrrri

**Reply by kangz12345** [reply]
How about putting the testing modules (e.g., `tests.py` in your example) outside the package?
Is it able to exclude such testing codes when installing the package?

**Reply by kangz12345** [reply]
Example file structure:
```
swift (repo)
- swift
- apps
- tests
  - test_swift.py
  - test_apps.py
```

**Reply by BECATRUE** [reply]
Looking for it, I agree with you since we can run unittest for a module as a whole!

**Reply by BECATRUE** [reply]
Is there any suggestion?

**Comment by BECATRUE** [comment]
Since there is no more suggestion, I decided testing structure in #88.
